src/v/kafka/protocol/tests/batch_reader_test.cc

```diff
@@ -31,3 +31,32 @@ SEASTAR_THREAD_TEST_CASE(batch_reader_last_offset) {
 
     BOOST_REQUIRE_EQUAL(crs.last_offset(), expected_last_offset);
 }
+
+SEASTAR_THREAD_TEST_CASE(consumer_records_consume_record_batch) {
+    constexpr model::offset init_offset{0};
+
+    // Create some batches
+    auto input = storage::test::make_random_batches(init_offset, 40);
+    BOOST_REQUIRE(!input.empty());
+    const auto expected_last_offset = input.back().last_offset();
+
+    // Serialise the batches
+    auto mem_res = model::make_memory_record_batch_reader(std::move(input))
+                     .consume(
+                       kafka::kafka_batch_serializer{}, model::no_timeout)
+                     .get();
+
+    // Functionality under test
+    auto crs = kafka::batch_reader(std::move(mem_res.data));
+
+    model::offset last_offset{};
+    while (!crs.empty()) {
+        auto kba = crs.consume_batch();
+        BOOST_REQUIRE(kba.v2_format);
+        BOOST_REQUIRE(kba.valid_crc);
+        BOOST_REQUIRE(kba.batch);
+        last_offset = kba.batch->last_offset();
+    }
+
+    BOOST_REQUIRE_EQUAL(last_offset, expected_last_offset);
```

Comment:
Maybe it would also be useful to assert the number of bytes processed/read against the expected value